###  Метод Монте-Карло  <br>

####  Расщепление набора данных на обучающее и тестовое множество

In [1]:
import pandas as pd
import numpy as np

import matplotlib
import matplotlib.pyplot as plt

#  matplotlib.style.use('ggplot')
#  Эта строчка нужна для того, чтобы картинки отображались в ячейках
%matplotlib inline


In [2]:
import random


Отрабатываем технику выбора части набора данных

In [3]:
#  Данные для иллюстрации. Шаг 1 

x = np.array(range(1,11))

print(x)


[ 1  2  3  4  5  6  7  8  9 10]


In [4]:
#   Выбор подмножества по номерам элементов

num = [2,7,3]

print(x[num])

array([3, 8, 4])

In [5]:
#   Выбор подмножества элементов, для которых True

num2 = np.array([True, False, True, False, True, False, True, False, True, False])

x[num2]

array([1, 3, 5, 7, 9])

In [7]:
#  Данные для иллюстрации. Шаг 2 

import os
os.chdir("C:/Users/USER/Documents/Python/_AD_Python_part_1_Lectures/AD_Python_1_02_cluster/1_beverage/")
df = pd.read_csv("beverage_r.csv", sep=";", index_col='numb.obs')
df.head()

COKE  D_COKE  D_PEPSI  D_7UP  PEPSI  SPRITE  TAB  SEVENUP
numb.obs                                                           
1            1       0        0      0      1       1    0        1
2            1       0        0      0      1       0    0        0
3            1       0        0      0      1       0    0        0
4            0       1        0      1      0       0    1        0
5            1       0        0      0      1       0    0        0

In [8]:
len(df)

34

In [9]:
num3 = np.array([True, False]*17)

num3

array([ True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False,  True, False,
        True, False,  True, False,  True, False,  True, False,  True,
       False,  True, False,  True, False,  True, False])

In [10]:
#   Выбор подмножества элементов, для которых True

df.iloc[num3, :]

COKE  D_COKE  D_PEPSI  D_7UP  PEPSI  SPRITE  TAB  SEVENUP
numb.obs                                                           
1            1       0        0      0      1       1    0        1
3            1       0        0      0      1       0    0        0
5            1       0        0      0      1       0    0        0
7            0       1        1      1      0       0    1        0
9            1       1        0      0      0       1    1        1
11           1       0        0      0      1       1    0        0
13           0       0        1      1      0       1    0        1
15           0       1        1      0      0       0    1        0
17           0       1        0      0      0       1    0        0
19           1       0        0      0      0       0    0        1
21           1       0        0      0      1       0    0        0
23           0       1        0      1      0       0    1        0
25           0       1        1      1      0       0    0        0
27           0       1        0      0      0       0    1        0
29           1       0        0      0      0       1    0        0
31           1       0        0      0      1       0    0        1
33           1       0        0      0      1       0    0        1

In [12]:
#  Зерно датчика случайных чисел
np.random.seed(seed=123456)


r1 = np.random.uniform(low = 0.0, high=1.0, size=len(df))
r1

array([0.12696983, 0.96671784, 0.26047601, 0.89723652, 0.37674972,
       0.33622174, 0.45137647, 0.84025508, 0.12310214, 0.5430262 ,
       0.37301223, 0.44799682, 0.12944068, 0.85987871, 0.82038836,
       0.35205354, 0.2288873 , 0.77678375, 0.59478359, 0.13755356,
       0.85289978, 0.23550748, 0.14622723, 0.58986877, 0.57401177,
       0.06126996, 0.590426  , 0.24534982, 0.34044494, 0.98472874,
       0.9195404 , 0.03777169, 0.86154929, 0.75356885])

In [13]:
#  Откуда появляются True  False?
#  В результате проверки условия отбора!
#  Выбираем те элементы, для которых случайное число > 0.5

r2 = r1 > 0.5
r2

array([False,  True, False,  True, False, False, False,  True, False,
        True, False, False, False,  True,  True, False, False,  True,
        True, False,  True, False, False,  True,  True, False,  True,
       False, False,  True,  True, False,  True,  True])

In [14]:
r1[r2]

array([0.96671784, 0.89723652, 0.84025508, 0.5430262 , 0.85987871,
       0.82038836, 0.77678375, 0.59478359, 0.85289978, 0.58986877,
       0.57401177, 0.590426  , 0.98472874, 0.9195404 , 0.86154929,
       0.75356885])

In [15]:
#  Выбор в одну строку, без создания r2
#  Хотя r2 все равно создается невидимо для нас, 
#  как служебный вектор

r1[r1 > 0.5]

array([0.96671784, 0.89723652, 0.84025508, 0.5430262 , 0.85987871,
       0.82038836, 0.77678375, 0.59478359, 0.85289978, 0.58986877,
       0.57401177, 0.590426  , 0.98472874, 0.9195404 , 0.86154929,
       0.75356885])

In [16]:
#  Всегда выбирать половину элемнтов вектора не получается,
#  возможны случайные отклонения


len(r1[r2])

16

In [17]:
#  Выберем остальные элементы, заменяя True на False и наоборот

r1[np.logical_not(r1 > 0.5)]

array([0.12696983, 0.26047601, 0.37674972, 0.33622174, 0.45137647,
       0.12310214, 0.37301223, 0.44799682, 0.12944068, 0.35205354,
       0.2288873 , 0.13755356, 0.23550748, 0.14622723, 0.06126996,
       0.24534982, 0.34044494, 0.03777169])

In [18]:
#  Сколько остальных элементов?
len(r1[np.logical_not(r1 > 0.5)])

18

In [35]:
#  Чтобы точнее разделить вектор на две равные части (почти равные)
#  используем квантили

np.quantile(r1, 0.5)

0.44968664758067156

In [19]:
r1[r1 > np.quantile(r1, 0.5)]

array([0.96671784, 0.89723652, 0.45137647, 0.84025508, 0.5430262 ,
       0.85987871, 0.82038836, 0.77678375, 0.59478359, 0.85289978,
       0.58986877, 0.57401177, 0.590426  , 0.98472874, 0.9195404 ,
       0.86154929, 0.75356885])

In [20]:
len(r1[r1 > np.quantile(r1, 0.5)])

17

In [21]:
len(r1[np.logical_not(r1 > np.quantile(r1, 0.5))])

17

Результат: мы умеем выбирать случайно подмножество элементов <br>
Далее в качестве элементов будем использовать строки таблица данных

In [23]:
#  Пример, как можно получить обучающее множество из 50% наблюдений 

df.iloc[r1 > np.quantile(r1, 0.5), :]

COKE  D_COKE  D_PEPSI  D_7UP  PEPSI  SPRITE  TAB  SEVENUP
numb.obs                                                           
2            1       0        0      0      1       0    0        0
4            0       1        0      1      0       0    1        0
7            0       1        1      1      0       0    1        0
8            1       1        0      0      1       1    0        1
10           1       0        0      0      1       0    0        1
14           1       0        0      0      0       1    0        0
15           0       1        1      0      0       0    1        0
18           1       1        0      0      1       0    0        0
19           1       0        0      0      0       0    0        1
21           1       0        0      0      1       0    0        0
24           1       1        0      0      1       0    0        0
25           0       1        1      1      0       0    0        0
27           0       1        0      0      0       0    1        0
30           0       1        1      0      0       0    1        0
31           1       0        0      0      1       0    0        1
33           1       0        0      0      1       0    0        1
34           0       1        1      1      0       0    1        0

In [24]:
#  Остальные 50% наблюдений составят тестовое множество. 

df.iloc[np.logical_not(r1 > np.quantile(r1, 0.5)), :]

COKE  D_COKE  D_PEPSI  D_7UP  PEPSI  SPRITE  TAB  SEVENUP
numb.obs                                                           
1            1       0        0      0      1       1    0        1
3            1       0        0      0      1       0    0        0
5            1       0        0      0      1       0    0        0
6            1       0        0      0      1       1    0        0
9            1       1        0      0      0       1    1        1
11           1       0        0      0      1       1    0        0
12           0       1        0      0      0       0    1        0
13           0       0        1      1      0       1    0        1
16           0       0        0      0      1       1    0        0
17           0       1        0      0      0       1    0        0
20           1       1        1      0      1       0    0        0
22           1       0        0      0      1       0    0        0
23           0       1        0      1      0       0    1        0
26           0       1        0      1      0       0    1        0
28           1       0        0      0      0       1    0        1
29           1       0        0      0      0       1    0        0
32           0       1        1      0      0       0    1        0

Задача 1. <br>
Разделить таблицу данных так, чтобы  <br>
в обучающую выборку попало 66% наблюдений,  <br>
а в тестовую выборку 33% наблюдений

Задача 2. <br>
Разделить таблицу данных на четыре одинаковых части.

Задача 3. <br>
Разделить таблицу данных на две стратифицированные подвыборки. <br>
По переменным кока-кола и минералка.